compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -0,0 +1,622 @@
+    : public OpRewritePattern<linalg::BatchMatmulOp> {
+public:
+  using OpRewritePattern<linalg::BatchMatmulOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::BatchMatmulOp bmmOp,
+                                PatternRewriter &rewriter) const override {
+    Value lhs = bmmOp.getInputs()[0];
+    Value rhs = bmmOp.getInputs()[1];
+    SmallVector<NamedAttribute> attrs = getPrunedAttributeList(bmmOp);
+    if (auto rhsTranspose = rhs.getDefiningOp<linalg::TransposeOp>()) {
+      if (arePermutationsEqual(rhsTranspose.getPermutation(), {0, 2, 1})) {
+        rewriter.replaceOpWithNewOp<linalg::BatchMatmulTransposeBOp>(
+            bmmOp, ValueRange{lhs, rhsTranspose.getInput()},
+            bmmOp.getDpsInits(), attrs);
+        return success();
+      }
+    }
+
+    if (auto lhsTranspose = lhs.getDefiningOp<linalg::TransposeOp>()) {
+      if (arePermutationsEqual(lhsTranspose.getPermutation(), {0, 2, 1})) {
+        rewriter.replaceOpWithNewOp<linalg::BatchMatmulTransposeAOp>(
+            bmmOp, ValueRange{lhsTranspose.getInput(), rhs},
+            bmmOp.getDpsInits(), attrs);
+        return success();
+      }
+    }
+    return failure();
+  }
+};
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Pass definition
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct PropagateLinalgTransposePass
+    : public PropagateLinalgTransposeBase<PropagateLinalgTransposePass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect, tensor::TensorDialect>();
+  }
+  PropagateLinalgTransposePass(bool enableAggressivePropagation) {
+    this->enableAggressivePropagation = enableAggressivePropagation;
+  }
+  PropagateLinalgTransposePass(const PropagateLinalgTransposePass &pass)
+      : PropagateLinalgTransposePass(pass.enableAggressivePropagation) {}
+
+  void runOnOperation() override;
+
+private:
+  Option<bool> keepTransposes{
+      *this, "keep-transposes",
+      llvm::cl::desc(
+          "Flag used for lit-testing without re-generalizing the transposes. "
+          "Not for general usage"),
+      llvm::cl::init(false)};
+
+  Option<bool> testSinkingOnly{
+      *this, "test-sinking-only",
+      llvm::cl::desc("Flag used for lit-testing sinking patterns only. "
+                     "Not for general usage"),
+      llvm::cl::init(false)};
+};
+} // namespace
+
+void PropagateLinalgTransposePass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  auto funcOp = getOperation();
+  // First, specialize all transposes to `linalg.transpose`. This dramatically
+  // simplifies all subsequent propagation patterns, both in matching and
+  // rewriting.
+  {
+    SmallVector<linalg::GenericOp> genericCandidates;
+    funcOp.walk([&](linalg::GenericOp genericOp) {
+      if (IREE::Flow::isNonNullAndOutsideDispatch(genericOp)) {
```

Comment:
This check needs to probably happen in all patterns above. If any of the producer or consumer is in a dispatch, we need to abort.